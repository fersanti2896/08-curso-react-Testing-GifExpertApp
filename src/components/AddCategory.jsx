import { PropTypes } from "prop-types";
import { useState } from "react"

export const AddCategory = ( { newCategory } ) => {
    const [inputValue, setInputValue] = useState('');
    
    /**
     * Función que obtiene el valor del input.
     * @param {event} event.target 
     */
    const onInputChange = ({ target }) => {
        setInputValue(target.value);
    }

    /**
     * Función que agrega el valor del input al arreglo de categorías.
     * @param {event} event 
     * @returns 
     */
    const onSubmit = (event) => {
        event.preventDefault();

        if( inputValue.trim().length <= 1 ) return;

        setInputValue('');
        newCategory(inputValue.trim()); 
    }

    return (
        <>
            <form onSubmit={ (event) => onSubmit(event) } aria-label="form">
                <input type="text"
                    placeholder="Buscar Gifs"
                    value={ inputValue }
                    onChange={ onInputChange }
                />
            </form>
        </>
    )
}

AddCategory.propTypes = {
    newCategory: PropTypes.func.isRequired
}