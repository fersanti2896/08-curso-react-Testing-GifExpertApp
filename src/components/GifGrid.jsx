import PropTypes from 'prop-types';

import { GifGridItem } from './GifGridItem';
import { useFetchGifs } from '../hooks/useFetchGifs';

export const GifGrid = ({ category }) => {
    const { images, isLoading } = useFetchGifs( category );
    
    return (
        <>
            <h5>{ category }</h5>
            { isLoading && (<h5>Cargando...</h5>) }

            <div className='card-grid'>
                { 
                    images.map( (image) => ( 
                        <GifGridItem key={ image.id } 
                                     title={ image.title }
                                     url={ image.url } /> 
                    ) ) 
                }
            </div>
        </>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}

GifGrid.defaultProps = {
    category: 'Shrek'
}
