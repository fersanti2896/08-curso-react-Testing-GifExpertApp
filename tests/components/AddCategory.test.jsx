import { fireEvent, render, screen } from '@testing-library/react';
import { AddCategory } from '../../src/components/AddCategory';

describe('Pruebas en <AddCategory />', () => {
    const inputValue = 'Shrek';
    const newCategory = jest.fn();

    test('Debe hacer match con el snapshot.', () => {
        const { container } = render( <AddCategory newCategory={ newCategory }/> );

        expect( container ).toMatchSnapshot();
    });

    test('Debe de cambiar el valor de la caja de texto.', () => { 
        render( <AddCategory newCategory={ () => {} }/> );

        const input = screen.getByRole('textbox');
        
        fireEvent.input( input, { target: { value: inputValue } } );
        expect( input.value ).toBe(inputValue);
    });

    test('Debe de llamar newCategory si el input tiene el valor.', () => {       
        render( <AddCategory newCategory={ newCategory }/> );

        const input = screen.getByRole('textbox');
        const form = screen.getByRole('form');

        fireEvent.input( input, { target: { value: inputValue } } );
        fireEvent.submit( form );

        expect( input.value ).toBe('');
        expect( newCategory ).toHaveBeenCalled();
        expect( newCategory ).toHaveBeenCalledTimes(1);
        expect( newCategory ).toHaveBeenCalledWith( inputValue );
    });

    test('No debe de llamar el newCategory si el input está vacío.', () => { 
        const newCategory = jest.fn();
        render( <AddCategory newCategory={ newCategory }/> );

        const form = screen.getByRole('form');
        
        fireEvent.submit( form );
        expect( newCategory ).toHaveBeenCalledTimes(0);
    });
});