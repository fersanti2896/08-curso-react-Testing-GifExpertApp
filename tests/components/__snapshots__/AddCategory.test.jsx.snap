// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AddCategory /> Debe hacer match con el snapshot. 1`] = `
<div>
  <form
    aria-label="form"
  >
    <input
      placeholder="Buscar Gifs"
      type="text"
      value=""
    />
  </form>
</div>
`;
